### PR TITLE
feat: switch twitter logo

### PR DESCRIPTION
**Pull request by lelenei**
## Description

Closes # mf-4632

## Type of change

- [ ] Documentation
- [ ] Code refactoring (Restructuring existing code w/o changing its observable behavior)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (a fix or feature that would make something no longer possible to do/require old user must upgrade their Mask Network to this new version)

## Checklist

- [ ] My code follows the style guidelines of this project.
- [ ] I have performed a self-review of my own code.
  - [ ] I have removed all in development `console.log`s
  - [ ] I have removed all commented code.
- [ ] I have commented on my code, particularly in hard-to-understand areas.
- [ ] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

If this PR depends on external APIs:

- [ ] I have configured those APIs with CORS headers to let extension requests get passed. 
  - chrome extension: `chrome-extension://[id]`
  - firefox extension: `moz-extension://[id]`
- [ ] I have delegated all web requests to the background service via the internal RPC bridge.

### CODE REVIEW

File: packages/plugins/SwitchLogo/src/SNSAdaptor/SwitchLogo.tsx [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+        flexDirection: 'row',
+        paddingTop: 16,
+        justifyContent: 'end',
+        alignItems: 'center',
+    },
+}))
+
+interface SwitchLogoProps {}
+
+export const SwitchLogo = memo<SwitchLogoProps>(() => {
```
Comment (guanbinrui) [inline packages/plugins/SwitchLogo/src/SNSAdaptor/SwitchLogo.tsx]:
```suggestion
export const SwitchLogoDialog = memo<SwitchLogoProps>(() => {
```

File: packages/types/src/SocialNetworkUI.ts [hunk 2]
```diff
@@ -102,6 +102,8 @@ export namespace SocialNetworkUI {
             tips?(signal: AbortSignal): void
             lens?(signal: AbortSignal): void
             profileCard?(signal: AbortSignal): void
+            /** Inject a logo */
+            logo?(signal: AbortSignal): void
```
Comment (guanbinrui) [inline packages/types/src/SocialNetworkUI.ts]:
```suggestion
```

File: packages/plugins/SwitchLogo/src/base.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+    description: { fallback: PLUGIN_DESCRIPTION },
+    publisher: { name: { fallback: '' }, link: '' },
+    enableRequirement: {
+        networks: {
+            type: 'opt-in',
+            networks: {
+                [CurrentSNSNetwork.Twitter]: true,
+            },
+        },
+        target: 'beta',
```
Comment (guanbinrui) [inline packages/plugins/SwitchLogo/src/base.ts]:
```suggestion
        target: 'stable',
```

File: packages/plugins/SwitchLogo/src/SNSAdaptor/index.tsx [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    },
+    ApplicationEntries: [
+        {
+            ApplicationEntryID: PLUGIN_ID,
+            appBoardSortingDefaultPriority: 10,
+            icon: <Icons.Twitter3 size={36} />,
+            name: <Trans ns={PluginID.SwitchLogo} i18nKey="title" />,
+            category: 'dapp',
+            recommendFeature,
+        },
```
Comment (guanbinrui) [inline packages/plugins/SwitchLogo/src/SNSAdaptor/index.tsx]:
```suggestion
```

File: packages/plugins/SwitchLogo/src/SNSAdaptor/index.tsx [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+const recommendFeature = {
+    description: <Trans i18nKey="title" ns={PluginID.SwitchLogo} />,
+    backgroundGradient: 'linear-gradient(360deg, #FFECD2 -0.43%, #FCB69F 99.57%)',
+}
+
+const sns: Plugin.SNSAdaptor.Definition = {
+    ...base,
+    init(signal, context) {},
+    GlobalInjection() {
+        return <SwitchLogo />
```
Comment (guanbinrui) [inline packages/plugins/SwitchLogo/src/SNSAdaptor/index.tsx]:
```suggestion
        return (
            <Modals />
            <LogoSwitcher />
        )
```

File: packages/shared/src/UI/modals/ApplicationBoardModal/ApplicationSettingPluginSwitch.tsx [hunk 6]
```diff
@@ -111,6 +111,8 @@ export const ApplicationSettingPluginSwitch = memo(function ApplicationSettingPl
         async (id: string, checked: boolean) => {
             if (id === PluginID.GoPlusSecurity && checked === false)
                 return CrossIsolationMessages.events.checkSecurityConfirmationDialogEvent.sendToAll({ open: true })
+            if (id === PluginID.SwitchLogo)
+                CrossIsolationMessages.events.checkSwtichLogoEvent.sendToAll({ disabled: !checked })
```
Comment (guanbinrui) [inline packages/shared/src/UI/modals/ApplicationBoardModal/ApplicationSettingPluginSwitch.tsx]:
```suggestion
```

File: packages/shared-base/src/types/SwitchLogo.ts [hunk 7]
```diff
@@ -0,0 +1,4 @@
+export enum SwitchLogoType {
```
Comment (guanbinrui) [inline packages/shared-base/src/types/SwitchLogo.ts:1]: Move it back to the plugin itself.

File: packages/shared-base/src/Messages/Events.ts [hunk 8]
```diff
@@ -261,3 +265,7 @@ export interface PopupWalletConnectEvent {
     open: boolean
     uri?: string
 }
+
+export interface SwitchTwitterLogoEvent {
```
Comment (guanbinrui) [inline packages/shared-base/src/Messages/Events.ts]: Use SingletonModal stead.

File: packages/mask/src/social-network-adaptor/twitter.com/injection/logo.tsx [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+        if (logo) {
+            // eslint-disable-next-line @masknet/browser-no-set-html
+            logo.innerHTML = BlueBirdHTML
+        }
+    } else {
+        if (logo) {
+            // eslint-disable-next-line @masknet/browser-no-set-html
+            logo.innerHTML = newIconHtml || ''
+        }
+    }
```
Comment (guanbinrui) [inline packages/mask/src/social-network-adaptor/twitter.com/injection/logo.tsx]: Wrap it in `useEffect`. In the clean callback, recover the original one.

File: packages/mask/src/social-network-adaptor/twitter.com/injection/logo.tsx [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    const current = useLastRecognizedIdentity()
+    const logoType = useValueRef(switchLogoSettings[current.identifier?.userId || ''])
+    const logo = logoSelector().evaluate()
+
+    const { value: switchLogoDisabled } = useAsync(async () => {
+        const disabled = await Services.Settings.getPluginMinimalModeEnabled(PluginID.SwitchLogo)
+        return disabled === BooleanPreference.True
+    }, [])
+
+    const [disabled, setDisabled] = useState(switchLogoDisabled)
```
Comment (guanbinrui) [inline packages/mask/src/social-network-adaptor/twitter.com/injection/logo.tsx]: Use `<GlobalInjection />` stead.

File: packages/mask/src/social-network-adaptor/twitter.com/injection/logo.tsx [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    attachReactTreeWithContainer(watcher.firstDOMProxy.beforeShadow, { untilVisible: true, signal }).render(<Logo />)
+}
+
+const onOpenDialog = () => {
+    CrossIsolationMessages.events.switchLogoUpdated.sendToLocal({ open: true })
+}
+
+function Logo() {
+    const current = useLastRecognizedIdentity()
+    const logoType = useValueRef(switchLogoSettings[current.identifier?.userId || ''])
```
Comment (guanbinrui) [inline packages/mask/src/social-network-adaptor/twitter.com/injection/logo.tsx]: 👍

**Review by Jack-Works — CHANGES_REQUESTED** [review]
please modify packages/plugins/tsconfig.json

File: packages/plugins/SwitchLogo/src/locales/zh-CN.json [hunk 12]
```diff
@@ -0,0 +1,4 @@
+{
+    "name": "示例插件",
```
Comment (Jack-Works) [inline packages/plugins/SwitchLogo/src/locales/zh-CN.json]: ?

File: packages/plugins/SwitchLogo/src/Dashboard/index.tsx [hunk 13]
```diff
@@ -0,0 +1,9 @@
+import type { Plugin } from '@masknet/plugin-infra'
+import { base } from '../base.js'
+
+const dashboard: Plugin.Dashboard.Definition = {
+    ...base,
+    init(signal, context) {},
+}
```
Comment (Jack-Works) [inline packages/plugins/SwitchLogo/src/Dashboard/index.tsx]: remove this file.

File: packages/plugins/SwitchLogo/src/SNSAdaptor/LogoSwitcher.tsx [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+        if (logo) {
+            // eslint-disable-next-line @masknet/browser-no-set-html
+            logo.innerHTML = BlueBirdHTML
+        }
+    } else {
+        if (logo) {
+            // eslint-disable-next-line @masknet/browser-no-set-html
+            logo.innerHTML = newIconHtml || ''
+        }
+    }
```
Comment (UncleBill) [inline packages/plugins/SwitchLogo/src/SNSAdaptor/LogoSwitcher.tsx]: Wrap with `useLayoutEffect`.

File: packages/shared-base/src/types/SwitchLogo.ts (hunk 7 above)
Comment (lelenei) [reply packages/shared-base/src/types/SwitchLogo.ts:1]: need by settings

**Pull request merged**